validphys2/src/validphys/asy_exponents.py

```diff
@@ -0,0 +1,266 @@
+        d_dx = FinDiff(2,dx,acc=4)
+        alphaGrid_values = -np.log(abs(pdfGrid_values))
+        alphaGrid_values = d_dx(alphaGrid_values)
+        alphaGrid_values[alphaGrid_values == -np.inf] = np.nan  # when PDF_i =0
+        alphaGrid = pdfGrid.copy_grid(grid_values=pdf.stats_class(alphaGrid_values))
+        
+    return alphaGrid
+
+@check_positive('Q')
+@make_argcheck(check_basis)
+@check_xlimits
+def beta_asy(pdf, *,
+             xmin: numbers.Real = 0.6,
+             xmax: numbers.Real = 0.9,
+             npoints: int = 100,
+             Q: numbers.Real = 1.65,
+             basis: (str, Basis),
+             flavours: (list, tuple, type(None)) = None):
+    """Returns a list of xplotting_grids containing the value of the asymptotic
+    exponent beta, as defined by the second relationship in Eq. (4) of 
+    [arXiv:1604.00024], at the specified value of Q (in GeV), in the interval [xmin, xmax].
+
+    basis: Is one of the bases defined in pdfbases.py. This includes 'flavour'
+    and 'evolution'.
+
+    flavours: A set of elements from the basis.
+    If None, the defaults for that basis will be selected.
+
+    npoints: the number of sub-intervals in the range [xmin, xmax] on which the 
+    derivative is computed.
+    """
+    checked = check_basis(basis, flavours)
+    basis = checked['basis']
+    flavours = checked['flavours']
+
+    if npoints == 2:
+        xGrid = np.array([xmin, xmax])
+    else:
+        xGrid = pdfgrids.xgrid(xmin, xmax, 'linear', npoints)
+
+
+    pdfGrid = pdfgrids.xplotting_grid(
+        pdf, Q, xgrid=xGrid, basis=basis, flavours=flavours)
+    pdfGrid_values = pdfGrid.grid_values.data
+    # NOTE: without this I get "setting an array element with a sequence"
+    xGrid = pdfGrid.xgrid
+    with warnings.catch_warnings():
+        warnings.simplefilter('ignore', RuntimeWarning)
+        dx = xGrid[1] - xGrid[0]
+        d_dx = FinDiff(2,dx,acc=4)
+        betaGrid_values = np.log(abs(pdfGrid_values))
+        betaGrid_values = (xGrid - 1.) * d_dx(betaGrid_values)
+        betaGrid_values[betaGrid_values == -np.inf] = np.nan  # when PDF_i =0
+        betaGrid = pdfGrid.copy_grid(grid_values=pdf.stats_class(betaGrid_values))
+
+    return betaGrid
+
+class AsyExponentPlotter(PDFPlotter):
+    """ Class inherenting from BandPDFPlotter, changing title and ylabel to reflect the effective
+    exponent being plotted.
+    """
+
+    def __init__(self, exponent, *args,  **kwargs):
+        self.exponent = exponent
+        super().__init__(*args, **kwargs)
+
+    def get_title(self, parton_name):
+        return fr"$\{self.exponent}_a$ for ${parton_name}$ at {format_number(self.Q, 3)} GeV"
+
+    def get_ylabel(self, parton_name):
+        if self.normalize_to is not None:
+            return "Ratio to {}".format(self.normalize_pdf.label)
+        else:
+            return fr"$\{self.exponent}_a$ for ${parton_name}$"
+
+class AsyExponentBandPlotter(BandPDFPlotter, AsyExponentPlotter):
```

Comment:
Any reason not to merge this into the class above?

Reply:
Merged.